src/AssocList.mo

```diff
@@ -120,6 +120,41 @@ module {
     rec(map)
   };
 
+  /// Remove the entry with `key` in `map`, by returning the reduced list.
+  /// Also returns the value of the removed entry if it existed and `null` otherwise.
+  /// Compares keys using the provided function `equal`.
+  ///
+  /// Example:
+  /// ```motoko include=import,initialize
+  /// // Add three entries to the map
+  /// // map = [(0, 10), (1, 11), (2, 12)]
+  /// map := AssocList.remove(map, 1, Nat.equal).0;
+  /// List.toArray(map)
+  /// ```
+  /// Runtime: O(size)
+  ///
+  /// Space: O(size)
+  ///
+  /// *Runtime and space assumes that `equal` runs in O(1) time and space.
+  public func remove<K, V>(
+    map : AssocList<K, V>,
+    key : K,
+    equal : (K, K) -> Bool
+  ) : (AssocList<K, V>, ?V) {
+    func rec(al : AssocList<K, V>) : (AssocList<K, V>, ?V) {
```

Comment:
This isn't tail recursive, so will fail on a large assoc lists, in case we care about that.

Reply:
The list size should be bound to 8 due to `MAX_LEAF_SIZE` in `Trie`.
Or do you see an easy way of changing it to tail recursion?

Reply:
You can introduce accumulating parameters and then reverse but that's expensive too, I guess. Maybe we shouldn't worry about it since our stacks are large and people should be punished for using AssocList on large maps